fabric/src/main/java/org/geysermc/hydraulic/fabric/FabricUtil.java

```diff
@@ -0,0 +1,188 @@
+package org.geysermc.hydraulic.fabric;
+
+import com.google.common.collect.ImmutableList;
+import com.google.gson.JsonArray;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonPrimitive;
+import net.minecraft.util.GsonHelper;
+import org.geysermc.hydraulic.fabric.platform.HydraulicFabricBootstrap;
+import org.geysermc.hydraulic.platform.mod.ModInfo;
+import org.jetbrains.annotations.Nullable;
+
+import java.io.IOException;
+import java.io.Reader;
+import java.nio.file.FileSystem;
+import java.nio.file.FileSystems;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Collection;
+import java.util.SortedMap;
+import java.util.TreeMap;
+import java.util.function.Consumer;
+import java.util.stream.Stream;
+
+/**
+ * Utilities used to implement {@link HydraulicFabricBootstrap}.
+ */
+public class FabricUtil {
+    public static final int ICON_SIZE = 256;
+
+    /**
+     * Creates {@link ModInfo} objects for each mod in a jar, including JiJ'd mods, recursively.
+     * @param roots The root paths of a mod. Usually only one element, but may be more if using classpath groups.
+     * @return A stream that returns each created {@link ModInfo} object.
+     * @see #resolveJiJ(Collection, Consumer)
+     */
+    public static Stream<ModInfo> resolveJiJ(Collection<Path> roots) {
+        return Stream.of(roots).mapMulti(FabricUtil::resolveJiJ);
+    }
+
+    /**
+     * Creates {@link ModInfo} objects for each mod in a jar, including JiJ'd mods, recursively.
+     * @param roots The root paths of a mod. Usually only one element, but may be more if using classpath groups.
+     * @param output Invoked with each created {@link ModInfo} object.
+     * @see #resolveJiJ(Collection)
+     */
+    // We aren't modifying jars, therefore we don't need to close() them, as they'll simply get GCed.
+    @SuppressWarnings("resource")
+    public static void resolveJiJ(Collection<Path> roots, Consumer<ModInfo> output) {
+        final Path fmjPath = resolveFile(roots, "fabric.mod.json");
+        if (fmjPath == null) return;
+
+        final JsonObject fmj;
+        try (Reader reader = Files.newBufferedReader(fmjPath)) {
+            fmj = GsonHelper.parse(reader);
+        } catch (IOException ignored) {
+            return;
```

Comment:
Should we not put a debug or warn here?

Reply:
Then a warning should be put in every place where an invalid jar is found, which is a very large number of places, and is a little unnecessary, as it can only happen if you're using a mod with Hydraulic that JiJs a mod which is client only and broken on the client.